Context deadline propagation should cascade

buchgr opened the pull request:
A call's timeout as specified in its metadata should be set depending
on the deadline of the call's context. If a call has an explicit deadline
set (through CallOptions), then the smaller deadline (from context and call options)
should be used to compute the timeout.

Minor Fixes:
 - A Context no longer uses a TimeoutException as the cancellation cause
   when its deadline has been exceeded, but a StatusException with the
   status set to Status.DEADLINE_EXCEEDED.
 - Introduced method Context.statusFromCancelledContext(Context)

louiscryan commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -164,6 +172,26 @@ public void start(final Listener<RespT> observer, Metadata headers) {
+    // timeouts.
+    final long now = System.nanoTime();
+    final Deadline callDeadline = callOptions.getDeadlineNanoTime() != null
+        ? Deadline.after(callOptions.getDeadlineNanoTime() - now, NANOSECONDS)
```
Consider implementing Deadline.at(instant, units) to avoid the unnecessary call to nanoTime just to compute the base.

buchgr replied:
@louiscryan yeah I was unsure about updating CallOptions. I ll update CallOptions then.

ejona86 replied:
Deadline.at(instant, units) is really strange; I don't think we really want that. Using Deadline in CallOptions is the way to go.

buchgr replied:
@louiscryan @ejona86 I assume you want the CallOptions change in a separate PR?

ejona86 replied:
It's fine to do in this PR, assuming it doesn't grow to be a thing unto itself (I was thinking it would be fairly small). Separate PR works fine as well.

louiscryan commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -164,6 +172,26 @@ public void start(final Listener<RespT> observer, Metadata headers) {
+
+    // TODO: Not nice. Improve. Maybe context should cancel itself immediately if deadline exceeded
+    if (context.getDeadline() != null && context.getDeadline().isExpired()) {
+      ((Context.CancellableContext) context).cancel(new TimeoutException("context timed out"));
```
Contexts do cancel immediately if the deadline is exceeded however the listeners are scheduled onto the executor

buchgr replied:
@louiscryan I might be missing something, but doesn't seem like Context gets canceled immediately if timeline is exceeded? See https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/Context.java#L659

louiscryan commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -198,7 +227,11 @@ public void runInContext() {
 
-    if (updateTimeoutHeader(callOptions.getDeadlineNanoTime(), headers)) {
+    final boolean deadlineExceeded = effectiveDeadline != null
+        && effectiveDeadline.deadlineNanos() - now <= 0;
```
effectiveDeadline.isExpired()

louiscryan commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -214,57 +247,51 @@ public void runInContext() {
+      headers.put(TIMEOUT_KEY, effectiveTimeout);
+    }
+
+    // Include both deadlines/timeouts for debugging purposes.
```
Why do we want these propagating on the wire? Wouldn't logging / recording in the created context satisfy the same need?

buchgr replied:
@louiscryan yes that's what I thought :-) but in the issue it says `Deadlines should be propagated explicitly in metadata for logging purposes ...` or am I misreading? Will revert to logging.

buchgr commented:
I am home for easter. will continue this on thursday.

buchgr commented:
I think I addressed all comments. Still missing are some tests and opt-out from propagation. I won't get to it before Monday, with that whole easter thing.

buchgr commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -215,45 +242,67 @@ public void runInContext() {
+  private static void updateTimeoutHeaders(@Nullable Deadline effectiveDeadline,
+      @Nullable Deadline callDeadline, @Nullable Deadline outerCallDeadline, Metadata headers) {
+    // TODO: Ask Louis: What if timeout is not specified in CallOptions but set in metadata?
     headers.removeAll(TIMEOUT_KEY);
```
@louiscryan if a timeout is set in the `Metadata` passed to `ClientCall.start(...)`, should that be ignored or should we use the minimum of `metadata timeout`, `calloptions deadline` and `context deadline` as the timeout?

ejona86 replied:
Generally we've taken the approach of always removing such metadata (independent of whether deadline is set elsewhere). So the removal here seems a Good Thing™ to me so that users are forced to use the correct API.

buchgr commented:
@louiscryan @ejona86  Sorry this took a while. Was on vacation with limited access to my laptop. 

PTAL. 

Only thing missing
```
Relatedly it should be simple for servers to opt out of deadline propagation when use-cases demand it.
```
Why does that make sense or what would be such a use case? If the context deadline is exceeded, the call will be canceled anyway?

buchgr commented on core/src/main/java/io/grpc/Context.java:
```diff
@@ -656,7 +655,7 @@ private CancellableContext(Context parent, Deadline deadline,
           @Override
           public void run() {
-            cancel(new TimeoutException("context timed out"));
+            cancel(Status.DEADLINE_EXCEEDED.withDescription("context timed out").asException());
```
I made this change so we don't have to special case for `TimeoutException` in a few places, but can use `Contexts.statusFromCancelledContext` everywhere. WDYT?

louiscryan replied:
Yeah, this seems cleaner

louiscryan commented on core/src/main/java/io/grpc/Contexts.java:
```diff
@@ -130,4 +132,21 @@ public void onReady() {
+   * Returns the {@link Status} of a cancelled context or {@code null} if the context
+   * is not cancelled.
+   */
+  public static Status statusFromCancelled(Context context) {
```
Consider moving this into Contexts so we can keep the separation I mentioned above (and support TimeoutException here)

louiscryan commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -215,45 +242,67 @@ public void runInContext() {
-    // even when returning false.
+  private static void updateTimeoutHeaders(@Nullable Deadline effectiveDeadline,
+      @Nullable Deadline callDeadline, @Nullable Deadline outerCallDeadline, Metadata headers) {
+    // TODO: Ask Louis: What if timeout is not specified in CallOptions but set in metadata?
```
Agree with Eric, can remove this todo

louiscryan commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -215,45 +242,67 @@ public void runInContext() {
+  private static void logIfContextNarrowedTimeout(long effectiveTimeout,
+      Deadline effectiveDeadline, @Nullable Deadline outerCallDeadline,
+      @Nullable Deadline callDeadline) {
+    if (!log.isLoggable(Level.INFO) && outerCallDeadline != effectiveDeadline) {
```
if (!log.isLoggable(Level.INFO) || outerCallDeadline > effectiveDeadline) {

louiscryan commented:
With regard to

"Relatedly it should be simple for servers to opt out of deadline propagation when use-cases demand it."

This is what Context.fork() is intended to solve and it should be possible for servers to use interceptors to apply that behavior uniformly should they so choose. Nothing that needs to be in this PR

buchgr commented:
@louiscryan PTAL

buchgr commented:
Test failure seems unrelated
```java
io.grpc.stub.StubConfigTest > testConfigureDeadlineNanoTime FAILED
    java.lang.AssertionError: 20000000 < 34851943
```

louiscryan commented on core/src/main/java/io/grpc/Contexts.java:
```diff
@@ -130,4 +134,24 @@ public void onReady() {
+      return null;
+    }
+    final Throwable cancellationCause = context.cancellationCause();
+    if (cancellationCause instanceof StatusException) {
```
Could be StatusRuntimeException too. Probably better to do something like

if (cancellationCause instanceof TimeoutException) {
  return Status.DEADLINE_EXCEEDED.withDescription(cancellationCause.getMessage());
} else if (cancellationCause == null) {
 return Status.CANCELLED;
} else {
 // Could check for UNKNOWN here and substitute in CANCELLED but probably overkill
 return Status.fromThrowable(cancellationCause);
}

louiscryan commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -165,19 +172,36 @@ public void start(final Listener<RespT> observer, Metadata headers) {
+      context = parentContext.withCancellation();
+    }
+
+    // TODO: Context should cancel itself immediately when expired, so we don't need this check.
```
Can you fix this in Context rather than here, no need to schedule the executor if its deadline already expired.

buchgr replied:
@louiscryan good point. I am actually not sure why I didn't fix it immediately :sweat_smile:

louiscryan commented:
Sorry for the lag. Just two smallish comments

buchgr commented:
@louiscryan PTAL. Good to squash and merge?

buchgr commented:
@louiscryan I am not sure why I didn't change the Context cancellation on expired deadline immediately ... seems stupid in retrospect :sweat_smile:

louiscryan commented:
LGTM. Good to commit

Thanks Jacob!

buchgr commented:
Thanks! Cherry-picked as fd8fd517d21165e1ad97a
